### PR TITLE
features/moderation

### PR DESCRIPTION
**Pull request by quinchs**
# Summary
This PR adds the base moderation commands and functions, as well as some useful services/handlers associated with the moderation code.

# Changes
### Interaction Service
The interaction service is designed to act like the interactivity service, it allows you to use `NextButtonAsync` and `NextSlashCommandAsync`.

### Moderation Handler
The moderation handler is the core of this PR, It's what takes in infractions and computes them. It's responsible for logging infractions, kicking/muting/banning users, mutes, and storing infractions.

### Moderation Module
This module contains all the moderation commands for CWE.
| Command name | Parameters                    | Description                                      |
|--------------|-------------------------------|--------------------------------------------------|
| warn         | `<@user> <reason>`            | Warns a user.                                    |
| kick         | `<@user> <reason>`            | Kicks a user from the CWE guild                  |
| ban          | `<@user> <reason>`            | Bans a user from the CWE guild                   |
| mute         | `<@user> <duration> <reason>` | Mutes a user for a given amount of time.         |
| infractions  | `<@user>`                     | Gets a list of the specified user's infractions. |

### RequreStaff Precondition Attribute
This attribute is designed for usage on commands, it extends the `PreconditionAttribute` class from Discord.Net.Commands. Attaching this attribute to your command will restrict its usage to only staff members of the CWE discord server.

# appsettings.json additions
This PR adds 3 new config fields

| Field location         | Type  | Description                                                                                               |
|------------------------|-------|-----------------------------------------------------------------------------------------------------------|
| Channels -> LogChannel | ulong | The channel to send moderation logs to.                                                                   |
| Roles -> Staff         | ulong | The staff role id. This Id is whats used to check whether or not a user can execute a moderation command. |
| Roles -> Muted         | ulong | The role to give when someone is muted with the `mute` command.                                           |

### CODE REVIEW

**Comment by n-Ultima** [comment]
Why make "Mute" it's own object? You can instead make an InfractionType Enum, or something like that. Now you have InfractionType.Warn, InfractionType.Mute, etc etc

File: CWE.Data/DataAccessLayer.cs [hunk 1]
```diff
@@ -311,6 +311,39 @@ public async Task DeleteSuggestion(int id)
...
+        /// <param name="infraction">The infraction to be created</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public async Task CreateInfraction(Infraction infraction)
+        {
+            if (this.dbContext.Infractions.Any(x => x.InfractionId == infraction.InfractionId))
+            {
+                return;
+            }
+
+            await this.dbContext.Infractions.AddAsync(infraction);
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:325]: In the DAL we don't use `AddAsync()`, please simply use `Add()`.

File: CWE.Data/DataAccessLayer.cs [hunk 2]
```diff
@@ -330,6 +363,50 @@ public async Task CreateRank(ulong id)
...
         /// <summary>
+        /// Gets all the infractions for a specified user.
+        /// </summary>
+        /// <param name="userId">The user to get the infractions for.</param>
+        /// <returns>A <see cref="IReadOnlyCollection{T}"/> containing all the infractions.</returns>
+        public async Task<IReadOnlyCollection<Infraction>> GetUserInfractions(ulong userId)
+        {
+            var infracs = this.dbContext.Infractions.Where(x => x.UserId == userId);
+            return await infracs.ToListAsync();
+        }
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:374]:
```suggestion
        public async Task<IReadOnlyCollection<Infraction>> GetInfractions(ulong userId)
        {
            return await this.dbContext.Infractions.Where(x => x.UserId == userId).ToListAsync();
        }
```
Can be moved into one line and changed the name as there is no `GetInfractions()` (more consistent naming).

File: CWE.Data/DataAccessLayer.cs [hunk 3]
```diff
@@ -330,6 +363,50 @@ public async Task CreateRank(ulong id)
...
+        /// <param name="mute">The mute to create</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public async Task CreateMute(Mute mute)
+        {
+            if (this.dbContext.Mutes.Any(x => mute.InfractionId == x.InfractionId))
+            {
+                return;
+            }
+
+            await this.dbContext.Mutes.AddAsync(mute);
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:388]: See the previous comment about the use of `AddAsync()`.

File: CWE.Data/DataAccessLayer.cs [hunk 4]
```diff
@@ -357,5 +434,15 @@ public async Task DeleteRank(ulong id)
...
         }
+
+        /// <summary>
+        /// Gets all current mutes.
+        /// </summary>
+        /// <returns>A list of the current mutes.</returns>
+        public async Task<List<Mute>> GetMutes()
+        {
+            return await this.dbContext.Mutes.ToListAsync();
+
```
Comment (hendrikdemir) [inline CWE.Data/DataAccessLayer.cs:445]: Unnecessary whitespace

File: CWE.Data/Models/Infraction.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        /// <summary>
+        /// Gets or sets the date this infraction was created on.
+        /// </summary>
+        public DateTime Date { get; set; }
+
+        /// <summary>
+        /// Gets or sets the reason this infraction was created.
+        /// </summary>
+        public string Reason { get; set; }
+    }
```
Comment (hendrikdemir) [inline CWE.Data/Models/Infraction.cs:77]: The usernames of members and staff members should not be stored as they can change over time and therefore cause confusion. Rather, only store the ids as properties `User` and `Staff`.

File: CWE.Data/Models/Mute.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+        /// <summary>
+        /// Gets or sets the time the mute started.
+        /// </summary>
+        public DateTime MuteStart { get; set; }
+
+        /// <summary>
+        /// Gets or sets the time the mute ends.
+        /// </summary>
+        public DateTime MuteEnd { get; set; }
```
Comment (hendrikdemir) [inline CWE.Data/Models/Mute.cs:35]:
```suggestion
        /// <summary>
        /// Gets or sets the time the mute started.
        /// </summary>
        public DateTime Start { get; set; }

        /// <summary>
        /// Gets or sets the time the mute ends.
        /// </summary>
        public DateTime End { get; set; }
```
Renaming of the properties as the class name is already mute, no need to prefix the properties with the name of the class they're in.

File: CWE.Data/Models/Mute.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    /// <summary>
+    /// Represents a mute structure to keep track of muted users.
+    /// </summary>
+    [Keyless]
+    public class Mute
+    {
+        /// <summary>
+        /// Gets or sets the infractio id for this mute.
+        /// </summary>
+        public Guid InfractionId { get; set; }
```
Comment (hendrikdemir) [inline CWE.Data/Models/Mute.cs:20]: Is there any reason this is called an `InfractionId`? I don't see a relationship to the infractions table, shouldn't this just be an id or `MuteId`?

File: CWE/Common/Extensions.cs [hunk 8]
```diff
@@ -61,5 +68,97 @@ public static string GetUsernameAndTag(this SocketUser socketUser)
...
+        /// <summary>
+        /// Makes a <see cref="EmbedAuthorBuilder"/> from a <see cref="IUser"/>.
+        /// </summary>
+        /// <param name="user">The user to make an <see cref="EmbedAuthorBuilder"/> from</param>
+        /// <returns>A <see cref="EmbedAuthorBuilder"/> whos properties are set to the current users.</returns>
+        public static EmbedAuthorBuilder GetAuthorEmbed(this IUser user)
+        {
+            return new EmbedAuthorBuilder()
+            {
+                Name = $"{user.Username}{user.Discriminator}",
```
Comment (hendrikdemir) [inline CWE/Common/Extensions.cs:81]:
```suggestion
                Name = $"{user.Username}#{user.Discriminator}",
```
Add the # between the username and discriminator.

**Review by hendrikdemir — CHANGES_REQUESTED** [review]
Thank you for this very useful PR! I would however like to see a bit more consistency with i.e. the creation of embeds. Where you sometimes use `.WithFields()`, in the original source we use `.AddField()`. Could you go over your code and ensure that both the methods that you use, as well as the syntax is consistent with the original code? This includes i.e. the use of colors (we have a colors class that includes a `success` color).

**Comment by n-Ultima** [comment]
Just because I'm bored and I decided to go over this, why exactly is the `Mute` model even there? Why not simply have a nullable `TimeSpan?` duration on the `Infraction` model? You can also simply add in a `DateTimeOffet?` for when the infraction should expire.

**Comment by n-Ultima** [comment]
That, and I see an `async void` method in `ModerationHandler.cs`. Won't that crash the program if any exception is thrown at all?

**Comment by hendrikdemir** [comment]
Hi @quinchs, do you have any updates to solve the comments?

**Comment by hendrikdemir** [comment]
Closed as there were no further updates, thank you for the PR :)

**Pull request closed without merging**